### PR TITLE
feature: introduce upgrade command

### LINKED ISSUE #325: Prototype upgrade command
Introduce upgrade command as discussed in [Part 1 of the design proposal](https://docs.google.com/document/d/1_iJJAASFYt4xdkprkmEkjSMEFX2_PfLrFSyKDnSC1Ag/edit?ts=600afe2d#).

Introduce a new CLI subcommand to initiate a cluster upgrade.

The upgrade command will only upgrade the cluster’s software with a “soft-upgrade” (for lack of a better word) strategy where the upgrade command will update the controller deployment with the new image version. After the new controller starts, it will update all the Kubernetes deployments, which will trigger the new versions’ redeployment.

If required, merge the previous controller configuration with any new settings introduced with the new version. This step is not required in this initial implementation. We only need to worry about it when new configuration options are introduced.

The command signature is:

```
usage: opstrace upgrade [-h] [--log-level LEVEL] [--yes] [--region REGION]
                       PROVIDER CLUSTER_NAME

positional arguments:
  PROVIDER              The cloud provider to act on (aws, gcp).
  CLUSTER_NAME          The Opstrace cluster name ([a-z0-9-_], no more than 13 characters).

optional arguments:
  -h, --help            show this help message and exit
  --log-level LEVEL     Set log level for output on stderr. One of debug, info, warning, error. Default: info
  --yes                 Automatic yes to prompts; assume 'yes' as the answer to all prompts and run non-interactively.
  --region REGION       Set the cluster AWS region. Only needed when the automatic region detection fails (when the corresponding EKS cluster cannot be found or inspected). Not yet supported for GCP
  --to-be-named         Deletes all the Opstrace components and reinstalls the controller.
```

The CLI version used to execute the upgrade command implicitly defines the upgrade target (the controller version). The intended method for initiating an upgrade is downloading a new CLI version and running it against an old cluster (a set up with an older CLI version).

Needs #324.

### PR DESCRIPTION
**Pull request by sreis**
Close #325

This PR adds the upgrade command to the opstrace CLI, which required a couple of refactorings:
* Adds build info to its own package to use that information between the different packages consistently.
* Moves functions to fetch the cluster kubeconfig to "./util" to share code between the commands.
* Moves functions to find the region to "./util" to share code between the commands.

### CODE REVIEW

**Comment by sreis** [comment]
I'm opening the PR now to check if I didn't break the tests.

File: packages/controller/src/tasks/general.ts [hunk 1]
```diff
@@ -34,5 +41,26 @@ export function* deployControllerResources(config: {
...
+      log.debug("creating controller resources");
+      yield all([resources.map(createResource)]);
+      break;
+    }
+
+    case ControllerResourcesDeploymentStrategy.Update: {
+      log.debug("updating controller resources");
+      yield all([resources.map(updateResource)]);
+      break;
+    }
```
Comment (sreis) [inline packages/controller/src/tasks/general.ts:60]: This is called but the resource is never updated...

```
2021-02-09T18:48:00.562Z debug: updating controller resources
2021-02-09T18:48:00.563Z debug: updateResource for hasura-admin-secret(ns: kube-system):
2021-02-09T18:48:00.563Z debug: updateResource for opstrace-controller(ns: kube-system):
2021-02-09T18:48:00.564Z debug: updateResource for opstrace-controller(ns: kube-system):
2021-02-09T18:48:00.564Z debug: updateResource for opstrace-controller-clusteradmin-binding(ns: default):
2021-02-09T18:48:00.565Z info: controller deployment found in informer-based state
```

**Comment by jgehrcke** [comment]
@sreis thanks for sharing early -- let me know when you'd like some feedback : ) 
Great progress. Feels good!

**Comment by sreis** [comment]
@jgehrcke Currently dealing with yarn dependency errors in a clean checkout. Have a look at the PR when you have spare time. Any review is more than welcome.

**Comment by sreis** [comment]
Fixed the build under a clean checkout and learned a bit about `tsconfig.json`. 

Fixed the issue with the controller resources not getting updated. The resource update request was not being sent because it's not waiting for the Promise to conclude. This is the same code pattern used in the installer and uninstaller. I'm not yet familiar with redux-saga to fully understand the issue. I worked around it by checking if the current deployment has been updated (this is done by waiting for the deployment status observed generation to change).

**Comment by triclambert** [comment]
Nit: update the copyright date.

**Comment by sreis** [comment]
Rebased with main. Going to merge when tests are green and follow up with additional PRs in case of post-merge reviews.

**Pull request merged**